File: src/main/java/com/codelog/lastlife/LifeContext.java
package com.codelog.lastlife;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;

import java.io.*;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.logging.Level;

public class LifeContext {
    private Map<UUID, Integer> lifeMap;
    private static LifeContext instance;

    private UUID bogeyman;

    private LifeContext() {
        lifeMap = new HashMap<>();
    }

    public static LifeContext get() {
        instance = (instance == null) ? new LifeContext() : instance;
        return instance;
    }

    public void setPlayerLives(UUID uuid, int lives) {
        lifeMap.put(uuid, lives);
        Player player = Bukkit.getServer().getPlayer(uuid);
        if (player == null)
            return;

        if (lives == 1) {
            Bukkit.getServer().dispatchCommand(Bukkit.getServer().getConsoleSender(),
                    "team join red " + player.getName());
        } else if (lives == 0) {
            Bukkit.getServer().dispatchCommand(Bukkit.getServer().getConsoleSender(),
                    "team leave " + player.getName());
            removePlayer(uuid);
        } else {
            Bukkit.getServer().dispatchCommand(Bukkit.getServer().getConsoleSender(),
                    "team join green " + player.getName());
        }
    }

    public int getPlayerLives(UUID p) {
        if (!lifeMap.containsKey(p))
            lifeMap.put(p, 0);
        return lifeMap.get(p);
    }

    public void removePlayer(UUID p) {
        lifeMap.remove(p);
    }

    public List<UUID> getAllPlayers() {
        return lifeMap.keySet().stream().toList();
    }

    public UUID getBogeyman() { return bogeyman; }
    public void setBogeyman(UUID p) {
        bogeyman = p;
    }

    public boolean hasPlayer(UUID p) { return lifeMap.containsKey(p); }

    public void writeToFile(File file) {
        if (file.exists()) {
            file.delete();
        }

        try {
            BufferedWriter writer = new BufferedWriter(new FileWriter(file));

            if (bogeyman != null) {
                writer.write(bogeyman.toString() + "\n");
            } else {

                writer.write("null\n");
            }

            for (UUID p : lifeMap.keySet()) {
                writer.write(p.toString() + "," + lifeMap.get(p) + "\n");
            }

            LastLife.logger.info("Done writing context file!");
            writer.close();
        } catch (IOException e) {
            LastLife.logger.log(Level.SEVERE, "Couldn't save context!");
        }
    }

    public void readFromFile(File file) {
        if (!file.exists()) {
            LastLife.logger.info("No context file, skipping...");
            return;
        }

        try {
            BufferedReader reader = new BufferedReader(new FileReader(file));

            String bogeymanStr = reader.readLine();
            if (bogeymanStr == null)
                return;

            if (!bogeymanStr.equals("null"))
                setBogeyman(UUID.fromString(bogeymanStr));

            String line;
            while ((line = reader.readLine()) != null) {

                int pos = line.indexOf(',');
                String uuid = line.substring(0, pos);
                String livesStr = line.substring(pos + 1);
                int lives = Integer.parseInt(livesStr);

                setPlayerLives(UUID.fromString(uuid), lives);
            }

            LastLife.logger.info("Done reading context file!");
        } catch (IOException e) {
            LastLife.logger.log(Level.SEVERE, "Couldn't read context file!");
        }
    }
}

File: src/main/kotlin/com/codelog/lastlife/LifeContext.kt
package com.codelog.lastlife

import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.io.File
import java.io.IOException
import java.util.*
import java.util.logging.Level
import kotlin.collections.HashMap

class LifeContext {

    companion object {
        var bogeyman: UUID? = null


        private val lifeMap: MutableMap<UUID, Int>

        init {
            lifeMap = HashMap()
        }

        fun setPlayerLives(uuid: UUID, lives: Int) {
            lifeMap[uuid] = lives
            val server = Bukkit.getServer()
            val player: Player = server.getPlayer(uuid) ?: return

            when (lives) {
                1 -> LastLife.redTeam?.addEntry(player.name)
                0 -> {
                    LastLife.redTeam?.removeEntry(player.name)
                    LastLife.greenTeam?.removeEntry(player.name)
                }
                else -> LastLife.greenTeam?.addEntry(player.name)
            }
        }

        fun getPlayerLives(uuid: UUID): Int {
            if (lifeMap[uuid] == null)
                lifeMap[uuid] = 0

            return lifeMap[uuid] ?: 0
        }

        fun removePlayer(p: UUID) {
            lifeMap.remove(p);
        }

        fun getAllPlayers(): List<UUID> {
            return lifeMap.keys.toList()
        }

        fun hasPlayer(uuid: UUID): Boolean = lifeMap.containsKey(uuid)

        fun writeToFile(file: File) {
            if (file.exists())
                file.delete()

            try {
                val writer = file.bufferedWriter()

                writer.write(bogeyman?.toString() ?: "null");
                writer.newLine()

                for (uuid in lifeMap.keys) {
                    val lives = lifeMap[uuid]
                    writer.write("$uuid,$lives")
                    writer.newLine()
                }

                LastLife.logger.info("Done writing context file!")
                writer.close()
            } catch (e: IOException) {
                LastLife.logger.log(Level.SEVERE, "Couldn't save context!")
            }
        }

        fun readFromFile(file: File) {
            if (!file.exists()) {
                LastLife.logger.warning("No context file, skipping...")
                return
            }

            try {
                val reader = file.bufferedReader()

                val bogeymanStr: String? = reader.readLine()
                if (bogeymanStr == null) {
                    LastLife.logger.log(Level.SEVERE, "Empty context file!")
                    return
                }

                if (bogeymanStr != "null") {
                    bogeyman = UUID.fromString(bogeymanStr)
                }

                for (line in reader.lineSequence()) {
                    val split = line.split(',')
                    setPlayerLives(UUID.fromString(split[0]), split[1].toIntOrNull() ?: 0)
                }

                LastLife.logger.info("Done reading context file!")
            } catch (e: IOException) {
                LastLife.logger.log(Level.SEVERE, "Couldn't read context file!")
            }
        }
    }
}